File: common/src/main/java/com/google/tsunami/common/net/http/HttpRequest.java
/*
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.tsunami.common.net.http;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import com.google.auto.value.AutoValue;
import com.google.common.base.Strings;
import com.google.errorprone.annotations.Immutable;
import com.google.protobuf.ByteString;
import java.util.Optional;
import okhttp3.HttpUrl;

/** Immutable HTTP request. */
@Immutable
@AutoValue
@AutoValue.CopyAnnotations
@SuppressWarnings("Immutable")
public abstract class HttpRequest {
  public abstract HttpMethod method();
  public abstract String url();
  public abstract HttpHeaders headers();
  public abstract Optional<ByteString> requestBody();

  public abstract Builder toBuilder();

  /**
   * Creates a {@link Builder} object for configuring {@link HttpRequest}.
   *
   * @return a {@link Builder} instance for the {@link HttpRequest} object.
   */
  public static Builder builder() {
    return new AutoValue_HttpRequest.Builder();
  }

  /**
   * Create a new HTTP GET request with the given {@code url}.
   *
   * @param url the url of the GET request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder get(String url) {
    checkArgument(!Strings.isNullOrEmpty(url));
    return builder().setMethod(HttpMethod.GET).setUrl(url);
  }

  /**
   * Create a new HTTP GET request with the given {@code uri}.
   *
   * @param uri the url of the GET request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder get(HttpUrl uri) {
    checkNotNull(uri);
    return builder().setMethod(HttpMethod.GET).setUrl(uri);
  }

  /**
   * Create a new HTTP HEAD request with the given {@code url}.
   *
   * @param url the url of the HEAD request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder head(String url) {
    checkArgument(!Strings.isNullOrEmpty(url));
    return builder().setMethod(HttpMethod.HEAD).setUrl(url);
  }

  /**
   * Create a new HTTP HEAD request with the given {@code uri}.
   *
   * @param uri the url of the HEAD request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder head(HttpUrl uri) {
    checkNotNull(uri);
    return builder().setMethod(HttpMethod.HEAD).setUrl(uri);
  }

  /**
   * Create a new HTTP POST request with the given {@code url}.
   *
   * @param url the url of the POST request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder post(String url) {
    checkArgument(!Strings.isNullOrEmpty(url));
    return builder().setMethod(HttpMethod.POST).setUrl(url);
  }

  /**
   * Create a new HTTP POST request with the given {@code uri}.
   *
   * @param uri the url of the POST request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder post(HttpUrl uri) {
    checkNotNull(uri);
    return builder().setMethod(HttpMethod.POST).setUrl(uri);
  }

  /**
   * Create a new HTTP PUT request with the given {@code url}.
   *
   * @param url the url of the PUT request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder put(String url) {
    checkArgument(!Strings.isNullOrEmpty(url));
    return put(HttpUrl.parse(url));
  }

  /**
   * Create a new HTTP PUT request with the given {@code uri}.
   *
   * @param uri the url of the PUT request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder put(HttpUrl uri) {
    checkNotNull(uri);
    return builder().setMethod(HttpMethod.PUT).setUrl(uri);
  }

  /**
   * Create a new HTTP DELETE request with the given {@code url}.
   *
   * @param url the url of the DELETE request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder delete(String url) {
    checkArgument(!Strings.isNullOrEmpty(url));
    return builder().setMethod(HttpMethod.DELETE).setUrl(url);
  }

  /**
   * Create a new HTTP DELETE request with the given {@code uri}.
   *
   * @param uri the url of the DELETE request.
   * @return a {@link Builder} object for configuring {@link HttpRequest}.
   */
  public static Builder delete(HttpUrl uri) {
    checkNotNull(uri);
    return builder().setMethod(HttpMethod.DELETE).setUrl(uri);
  }

  /** Builder for {@link HttpRequest}. */
  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setMethod(HttpMethod method);
    public abstract Builder setUrl(String url);
    public Builder setUrl(HttpUrl url) {
      setUrl(url.toString());
      return this;
    }
    public abstract Builder setHeaders(HttpHeaders httpHeaders);
    public abstract Builder setRequestBody(ByteString requestBody);
    public abstract Builder setRequestBody(Optional<ByteString> requestBody);

    public Builder withEmptyHeaders() {
      setHeaders(HttpHeaders.builder().build());
      return this;
    }

    abstract HttpRequest autoBuild();
    public HttpRequest build() {
      HttpRequest httpRequest = autoBuild();

      switch (httpRequest.method()) {
        case GET:
        case HEAD:
          checkState(
              !httpRequest.requestBody().isPresent(),
              "A request body is not allowed for HTTP GET/HEAD request");
          break;
        case POST:
        case PUT:
        case DELETE:
          break;
      }

      return httpRequest;
    }
  }
}

File: plugin_server/py/common/net/http/http_request.py
"""HTTP request utility."""

from typing import Optional

from tsunami.plugin_server.py.common.net.http.http_headers import Builder as HttpHeadersBuilder
from tsunami.plugin_server.py.common.net.http.http_headers import HttpHeaders
from tsunami.plugin_server.py.common.net.http.http_method import HttpMethod


def check_url_argument(func):
  def wrapper(cls, url):
    if not url:
      raise ValueError('Url cannot be None.')
    return func(cls, url)
  return wrapper


class HttpRequest:
  """HTTP request utility class.

  Please use Builder() to create instances of this class.

  Attributes:
    method: The HTTP request type.
    url: String address of the request.
    headers: The HTTP request headers in key/value pairs.
    body: The HTTP body could be empty per the request type. GET and
      HEAD request types must have empty request_body.
  """

  def __init__(self):
    self.method: HttpMethod = None
    self.url: Optional[str] = None
    self.headers: HttpHeaders = None
    self.body: Optional[bytes] = None

  @classmethod
  @check_url_argument
  def get(cls, url: str):
    return Builder().set_method(HttpMethod.GET).set_url(url)

  @classmethod
  @check_url_argument
  def head(cls, url: str):
    return Builder().set_method(HttpMethod.HEAD).set_url(url)

  @classmethod
  @check_url_argument
  def post(cls, url: str):
    return Builder().set_method(HttpMethod.POST).set_url(url)

  @classmethod
  @check_url_argument
  def put(cls, url: str):
    return Builder().set_method(HttpMethod.PUT).set_url(url)

  @classmethod
  @check_url_argument
  def delete(cls, url: str):
    return Builder().set_method(HttpMethod.DELETE).set_url(url)

  @classmethod
  def builder(cls):
    return Builder()


class Builder:
  """Builder class to create HTTP request object.

  Attributes:
    http_request: HTTP request object built.
  """

  def __init__(self):
    self.http_request = HttpRequest()

  def set_method(self, method: HttpMethod) -> 'Builder':
    """Set the HttpMethod type."""
    self.http_request.method = method
    return self

  def set_url(self, url: str) -> 'Builder':
    """Set the string address for the request."""
    self.http_request.url = url
    return self

  def set_headers(self, headers: HttpHeaders) -> 'Builder':
    """Set the HttpHeaders for the request."""
    self.http_request.headers = headers
    return self

  def set_request_body(self, request_body: Optional[bytes] = None) -> 'Builder':
    """Set the request body."""
    self.http_request.body = request_body
    return self

  def with_empty_headers(self) -> 'Builder':
    """Set an empty Http_headers for the request."""
    self.set_headers(HttpHeadersBuilder().build())
    return self

  def build(self) -> 'HttpRequest':
    if (
        self.http_request.method == HttpMethod.GET
        or self.http_request.method == HttpMethod.HEAD
    ):
      if self.http_request.body:
        raise ValueError(
            'A request body is not allowed for HTTP GET/HEAD request.')
    return self.http_request
